benchmark/atoi.cpp

```diff
@@ -0,0 +1,219 @@
+}
+
+// Note: eight digits can represent from 0 to (10^9) - 1, the logarithm base 2
+// of 10^9 is slightly less than 30, thus, only 30 bits are needed.
+auto lemire_as_zoo_swar(const char *chars) {
+    uint64_t bytes;
+    memcpy(&bytes, chars, 8);
+    auto allCharacterZero = zoo::meta::BitmaskMaker<uint64_t, '0', 8>::value;
+    using S8_64 = zoo::swar::SWAR<8, uint64_t>;
+    S8_64 convertedToIntegers = S8_64{bytes - allCharacterZero};
+    /* the idea is to perform the following multiplication:
+     * NOTE: THE BASE OF THE NUMBERS is 256 (2^8), then 65536 (2^16), 2^32
+     * convertedToIntegers is IN BASE 256 the number ABCDEFGH
+     * BASE256:   A    B    C    D    E    F    G    H *
+     * BASE256                                 10    1 =
+     *           --------------------------------------
+     * BASE256  1*A  1*B  1*C  1*D  1*E  1*F  1*G  1*H +
+     * BASE256 10*B 10*C 10*D 10*E 10*F 10*G 10*H    0
+     *           --------------------------------------
+     * BASE256 A+10B ....................... G+10H   H
+     * See that the odd-digits (base256) contain 10*odd + even
+     * Then, we can use base(2^16) digits, and base(2^32) to
+     * calculate the conversion for the digits in 100s and 10,000s
+     */
+    auto by11base256 = convertedToIntegers.multiply(256*10 + 1);
+    auto bytePairs = zoo::swar::doublePrecision(by11base256).odd;
+    static_assert(std::is_same_v<decltype(bytePairs), zoo::swar::SWAR<16, uint64_t>>);
+    auto by101base2to16 = bytePairs.multiply(1 + (100 << 16));
+    auto byteQuads = zoo::swar::doublePrecision(by101base2to16).odd;
+    auto by10001base2to32 = byteQuads.multiply(1 + (10000ull << 32));
+    return uint32_t(by10001base2to32.value() >> 32);
+}
+
+namespace zoo {
+
+std::size_t c_strLength(const char *s) {
+    using S = swar::SWAR<8, std::size_t>;
+    constexpr auto
+        MSBs = S{S::MostSignificantBit},
+        Ones = S{S::LeastSignificantBit};
+    S bytes;
+    for(auto base = s;; base += 8) {
+        memcpy(&bytes.m_v, base, 8);
+        auto firstNullTurnsOnMSB = bytes - Ones;
+        // The first lane with a null will borrow and set its MSB on when
+        // subtracted one.
+        // The borrowing from the first null interferes with the subsequent
+        // lanes, that's why we focus on the first null.
+        // The lanes previous to the first null might keep their MSB on after
+        // subtracting one (if their value is greater than 0x80).
+        // This provides a way to detect the first null: It is the first lane
+        // in firstNullTurnsOnMSB that "flipped on" its MSB
+        auto cheapestInversionOfMSBs = ~bytes;
+        auto firstMSBsOnIsFirstNull =
+            firstNullTurnsOnMSB & cheapestInversionOfMSBs;
+        auto onlyMSBs = zoo::swar::convertToBooleanSWAR(firstMSBsOnIsFirstNull);
+        if(onlyMSBs) { // there is a null!
+            auto firstNullIndex = onlyMSBs.lsbIndex();
+            return firstNullIndex + (base - s);
```

Comment:
Obviously this is wildly unsafe and will happily run through memory, just like the normal strlen.

Reply:
This is not unsafe.
If by safety we are going to mean 'it guards against any potential mistake', we will never be able to deliver, all code would be unsafe.
Indeed this also requires a null, just like the specification of `strlen`:
https://en.cppreference.com/w/cpp/string/byte/strlen
I prefer to represent strings using a variable length encoding of the size of the string, but alas, this is a demo to compare things to see the benefits of a library.  I've always thought the null-terminated strings to be problematic.

Reply:
It is as safe as strlen, no safer, no less safe.  Just commentary for color.
I also think null terminated strings are problematic.
We can ship something later that has a count.